app/src/main/java/com/github/onlynotesswent/model/folder/FolderRepositoryFirestore.kt

```diff
@@ -0,0 +1,147 @@
+        if (folder != null) {
+          onSuccess(folder)
+        } else {
+          onFailure(Exception("Folder not found"))
+        }
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun getFoldersFrom(
+      userId: String,
+      onSuccess: (List<Folder>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        val userFolders =
+            task.result.documents
+                .mapNotNull { document -> documentSnapshotToFolder(document) }
+                .filter { it.userId == userId }
+        onSuccess(userFolders)
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun getRootFoldersFrom(
+      userId: String,
+      onSuccess: (List<Folder>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        val rootFolders =
+            task.result.documents
+                .mapNotNull { document -> documentSnapshotToFolder(document) }
+                // Log.d("Firestore", "Fetched folder: ${folder?.name}, parentFolderId:
+                // ${folder?.parentFolderId}")
+                .filter { it.userId == userId && it.parentFolderId == null } // Only root folders
+        // Log.d("Firestore", "Filtered root folders: ${rootFolders.map { it.name }}")
+        onSuccess(rootFolders)
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun updateFolder(folder: Folder, onSuccess: () -> Unit, onFailure: (Exception) -> Unit) {
+    db.collection(collectionPath).document(folder.id).set(folder).addOnCompleteListener { result ->
+      if (result.isSuccessful) {
+        onSuccess()
+      } else {
+        result.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  override fun getSubFoldersOf(
+      parentFolderId: String,
+      onSuccess: (List<Folder>) -> Unit,
+      onFailure: (Exception) -> Unit
+  ) {
+    db.collection(collectionPath).get().addOnCompleteListener { task ->
+      if (task.isSuccessful) {
+        val userFolders =
+            task.result.documents
+                .mapNotNull { document -> documentSnapshotToFolder(document) }
+                .filter { it.parentFolderId == parentFolderId }
+        onSuccess(userFolders)
+      } else {
+        task.exception?.let { onFailure(it) }
+      }
+    }
+  }
+
+  /**
+   * Converts a DocumentSnapshot to a Folder object.
+   *
+   * @param document The DocumentSnapshot to convert.
+   * @return The converted Folder object.
+   */
+  fun documentSnapshotToFolder(document: DocumentSnapshot): Folder? {
+    return Folder(
+        id = document.id,
+        name = document.getString("name") ?: return null,
+        userId = document.getString("userId") ?: return null,
```

Comment:
It should never throw an error as name and userId will always be defined in Firestore. If it does, there probably is something wrong with Firestore, so I will make the function non-nullable and instead of returning null I will add !!